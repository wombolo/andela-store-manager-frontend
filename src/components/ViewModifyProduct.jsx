import React, { Component } from 'react';
import {getSingleProducts, editProduct, deleteProduct} from '../actions/productActions';
import PropTypes from "prop-types";
import {connect} from "react-redux";

export class ViewModifyProduct extends Component {
  state= {};

  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getSingleProducts(id);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.editProduct(this.state);
  };

  handleDeleteProduct = (id) => {
    if (confirm('Are you sure to delete')){
      this.props.deleteProduct(id);
    }
  };

  handleChange = (event) => {
    if (event.target.id === 'image_file') {
      const img = document.getElementById('image_file').files[0];
      this.setState({image_file: img});
    }
    else {
      this.setState({[event.target.id]: event.target.value});
    }
  };

  componentWillReceiveProps(nextProps) {
    if(nextProps.product) {
      const {product} = nextProps;
      this.setState(product);
    }
  }

  render() {
    const { profile } = this.props;
    if (profile.role === 'admin'){
      return (
         <div className="row clearfix products-grid">
            <div className="col-10 m-0-auto m-t-5">
              <h3>{this.state.title}</h3>
              <form onSubmit={this.handleSubmit}>

                <div className="row">

                  <div className="col-3">
                    <figure className="image-holder">
                      <img alt={'image'} src={`../assets/images/${this.state.image}`} />
                      <input type={'file'} id={'image_file'} onChange={this.handleChange}/>
                    </figure>
                  </div>


                  <div className="col-7">
                    <table border="0" width="100%" className="product-single-table table-striped">
                      <tbody>
                      <tr>
                        <td><label htmlFor="title">Name</label></td>
                        <td><input id="title" type="text" required={true} defaultValue={this.state.title} onChange={this.handleChange}/></td>
                      </tr>
                      <tr>
                        <td><label htmlFor="description">Description</label></td>
                        <td><textarea rows={15} id="description" required={true} onChange={this.handleChange}
                                      value={this.state.description}/></td>
                      </tr>

                      <tr>
                        <td><label htmlFor="price">Price</label></td>
                        <td><input id="price" type="number" step={'.01'}
                                   defaultValue={this.state.price}
                                   required={true} onChange={this.handleChange}/></td>
                      </tr>

                      <tr>
                        <td><label htmlFor="quantity">Quantity Available</label></td>
                        <td><input id="quantity" type={'number'}
                                   defaultValue={this.state.quantity}
                                   required={true} onChange={this.handleChange}/></td>
                      </tr>

                      <tr>
                        <td><label>Product Category</label></td>
                        <td><select>
                          <option>Fiction/ Sci-Fi</option>
                          <option>Comedy</option>
                          <option>Romance</option>
                          <option>Adventure</option>
                          <option>Thriller</option>
                          <option>Drama</option>
                        </select></td>
                      </tr>


                      <tr>
                        <td><label htmlFor="price">Creation Date</label></td>
                        <td>{ new Date(this.state.cdate).toDateString()}</td>
                      </tr>

                      <tr>
                        <td>&nbsp;</td>
                        <td><button type="submit" className="btn btn-primary">Update</button>
                          <button type='reset' onClick={() => (this.handleDeleteProduct(this.state.id))} className="btn btn-danger">Delete Product</button></td>
                      </tr>
                      </tbody>

                    </table>
                  </div>
                </div>
              </form>
            </div>
          </div>
      )
    }
    else {
      return (
        <div className="row clearfix products-grid">
          <div className="col-10 m-0-auto m-t-5">
            <h3>{this.state.title}</h3>

              <div className="row">
                <div className="col-3">
                  <figure className="image-holder">
                    <img alt={'image'} src={`../assets/images/${this.state.image}`} />
                  </figure>
                </div>

                <div className="col-7">
                  <table border="0" width="100%" className="product-single-table table-striped">
                    <tbody>
                    <tr>
                      <td><label htmlFor="title">Name</label></td>
                      <td>{this.state.title}</td>
                    </tr>
                    <tr>
                      <td><label htmlFor="description">Description</label></td>
                      <td>{this.state.description}</td>
                    </tr>

                    <tr>
                      <td><label htmlFor="price">Price</label></td>
                      <td>{this.state.price}</td>
                    </tr>

                    <tr>
                      <td><label htmlFor="quantity">Quantity Available</label></td>
                      <td>{this.state.quantity}</td>
                    </tr>

                    <tr>
                      <td><label>Product Category</label></td>
                      <td>Thriller</td>
                    </tr>

                    <tr>
                      <td><label htmlFor="price">Creation Date</label></td>
                      <td>{ new Date(this.state.cdate).toDateString()}</td>
                    </tr>

                    <tr>
                      <td>&nbsp;</td>
                      <td>
                        <button className="pull-right min-btn btn-primary" onClick={() => this.addToCart(this.state)}>Add to Cart</button>
                        </td>
                    </tr>

                    </tbody>

                  </table>
                </div>
              </div>
          </div>
        </div>
      )
    }

  }
}

ViewModifyProduct.propTypes = {
  getSingleProducts: PropTypes.func.isRequired
};


export const mapStateToProps = state => ({
  product: state.products.singleProduct,
  profile: state.auth.user.profile
});

export default connect(mapStateToProps, {
  getSingleProducts, editProduct, deleteProduct,
})(ViewModifyProduct);
