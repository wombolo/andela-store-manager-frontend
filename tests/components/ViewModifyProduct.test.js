import 'babel-polyfill'
import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { MemoryRouter } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

import { ViewModifyProduct } from '../../src/components/ViewModifyProduct.jsx';

global.confirm = () => true;


let wrapper;

const props = {
  match:{ params: 2},
  getSingleProducts: () => (
    {
      product: {id: 2, title: "Bonjour", image: "products/product-5.png",price: 150, quantity: 3, ViewModifyProductQty: '3', status: "active"}}
    ),

  addToCart: () => Promise.resolve(),
  editProduct: () => Promise.resolve(),
  deleteProduct: () => Promise.resolve(),
  product: {
    id: 2, title: "Bonjour", image: "products/product-5.png",price: 150, quantity: 3, ViewModifyProductQty: '3', status: "active"
  },

  profile: {
    id: 1, firstname: 'Akano', lastname: 'Peter',
    email: 'elkenzom1@gmail.com', role: 'admin', image: 'team-7.jpg',
  },
};


const mockStore = configureStore([thunk]);

const store = mockStore({ ...props });


describe('ViewModifyProduct Component', () => {
  it('should match snapshot', () => {
    wrapper = shallow(<ViewModifyProduct {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Methods in Class', () => {
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
          <ViewModifyProduct {...props} />
          </MemoryRouter>
        </Provider>,
    );
      // const ViewModifyProductComp = wrapper.find('ViewModifyProduct');

    });

    afterEach(() => {
      wrapper.unmount();
    });


    it('triggers handleSubmit() in ViewModifyProductComp', () => {
      const ViewModifyProductComp = wrapper.find('ViewModifyProduct');
      sinon.spy(ViewModifyProductComp.instance(), 'handleSubmit');

      ViewModifyProductComp.instance().handleSubmit({preventDefault: () => {}});
      expect(ViewModifyProductComp.instance().handleSubmit.calledOnce)
        .toEqual(true);
    });


    it('triggers handleChange() for input title and onClick() for button addtocart', () => {
      const wrapperNew = shallow(<ViewModifyProduct {...props} />);
      wrapperNew.setProps({product: props.product});

      const ViewModifyProductComp = wrapper.find('ViewModifyProduct');

      const file = {lastModified: 1547454659333,
        lastModifiedDate: 'Mon Jan 14 2019 09:30:59 GMT+0100 (West Africa Standard Time) {}',
        name: "15158524_xl.jpg",
        size: 3734559,
        type: "image/jpeg",
        webkitRelativePath: ""};

      const image_file = ViewModifyProductComp.find('#image_file').first();
      image_file.simulate('change', { target: { files: [file] , id: 'image_file'} });


      const buttonCartAdd = ViewModifyProductComp.find('button').last();
      buttonCartAdd.simulate('click',{});


      sinon.spy(ViewModifyProductComp.instance(), 'handleChange');
      ViewModifyProductComp.instance().handleChange({target: { value: 'u' , id: 'title'} });
      expect(ViewModifyProductComp.instance().handleChange.calledOnce)
        .toEqual(true);
    });


    it('triggers handleDeleteProduct() in ViewModifyProductComp', () => {

      const ViewModifyProductComp = wrapper.find('ViewModifyProduct');
      sinon.spy(ViewModifyProductComp.instance(), 'handleDeleteProduct');
      ViewModifyProductComp.instance().handleDeleteProduct(1);
      expect(ViewModifyProductComp.instance().handleDeleteProduct.calledOnce)
        .toEqual(true);
    });


    it('triggers handleSubmit() in ViewModifyProductComp for non-admin ', () => {
      props.profile.role='store_attendant';
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
            <ViewModifyProduct {...props} />
          </MemoryRouter>
        </Provider>,
      );

      const ViewModifyProductComp = wrapper.find('ViewModifyProduct');

      const buttonCartAdd = ViewModifyProductComp.find('button').last();
      buttonCartAdd.simulate('click',{});


      sinon.spy(ViewModifyProductComp.instance(), 'handleSubmit');

      ViewModifyProductComp.instance().handleSubmit({preventDefault: () => {}});
      expect(ViewModifyProductComp.instance().handleSubmit.calledOnce)
        .toEqual(true);
    });

  });
});
