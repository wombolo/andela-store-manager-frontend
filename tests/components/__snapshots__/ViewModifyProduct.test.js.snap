// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewModifyProduct Component should match snapshot 1`] = `
<div
  className="row clearfix products-grid"
>
  <div
    className="col-10 m-0-auto m-t-5"
  >
    <h3 />
    <form
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="col-3"
        >
          <figure
            className="image-holder"
          >
            <img
              alt="image"
              src=""
            />
            <input
              id="image_file"
              onChange={[Function]}
              type="file"
            />
          </figure>
        </div>
        <div
          className="col-7"
        >
          <table
            border="0"
            className="product-single-table table-striped"
            width="100%"
          >
            <tbody>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Name
                  </label>
                </td>
                <td>
                  <input
                    defaultValue=""
                    id="title"
                    onChange={[Function]}
                    required={true}
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="description"
                  >
                    Description
                  </label>
                </td>
                <td>
                  <textarea
                    id="description"
                    onChange={[Function]}
                    required={true}
                    rows={15}
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="price"
                  >
                    Price
                  </label>
                </td>
                <td>
                  <input
                    defaultValue={0}
                    id="price"
                    onChange={[Function]}
                    required={true}
                    step=".01"
                    type="number"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="quantity"
                  >
                    Quantity Available
                  </label>
                </td>
                <td>
                  <input
                    defaultValue={0}
                    id="quantity"
                    onChange={[Function]}
                    required={true}
                    type="number"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label>
                    Product Category
                  </label>
                </td>
                <td>
                  <select>
                    <option>
                      Fiction/ Sci-Fi
                    </option>
                    <option>
                      Comedy
                    </option>
                    <option>
                      Romance
                    </option>
                    <option>
                      Adventure
                    </option>
                    <option>
                      Thriller
                    </option>
                    <option>
                      Drama
                    </option>
                  </select>
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="price"
                  >
                    Creation Date
                  </label>
                </td>
                <td>
                  Invalid Date
                </td>
              </tr>
              <tr>
                <td>
                  Â 
                </td>
                <td>
                  <button
                    className="btn btn-primary"
                    type="submit"
                  >
                    Update
                  </button>
                  <button
                    className="btn btn-danger"
                    onClick={[Function]}
                    type="reset"
                  >
                    Delete Product
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </form>
  </div>
</div>
`;
